### PR TITLE
[SPARK-30125][SQL] Remove PostgreSQL dialect

### PR DESCRIPTION
**Pull request by xuanyuanking**
### What changes were proposed in this pull request?
Reprocess all PostgreSQL dialect related PRs, listing in order:

- #25158: PostgreSQL integral division support [revert]
- #25170: UT changes for the integral division support [revert]
- #25458: Accept "true", "yes", "1", "false", "no", "0", and unique prefixes as input and trim input for the boolean data type. [revert]
- #25697: Combine below 2 feature tags into "spark.sql.dialect" [revert]
- #26112: Date substraction support [keep the ANSI-compliant part]
- #26444: Rename config "spark.sql.ansi.enabled" to "spark.sql.dialect.spark.ansi.enabled" [revert]
- #26463: Cast to boolean support for PostgreSQL dialect [revert]
- #26584: Make the behavior of Postgre dialect independent of ansi mode config [keep the ANSI-compliant part]

### Why are the changes needed?
As the discussion in http://apache-spark-developers-list.1001551.n3.nabble.com/DISCUSS-PostgreSQL-dialect-td28417.html, we need to remove PostgreSQL dialect form code base for several reasons:
1. The current approach makes the codebase complicated and hard to maintain.
2. Fully migrating PostgreSQL workloads to Spark SQL is not our focus for now.

### Does this PR introduce any user-facing change?
Yes, the config `spark.sql.dialect` will be removed.

### How was this patch tested?
Existing UT.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
cc @gatorsmile

**Comment by xuanyuanking** [comment]
Thanks Dongjoon, also cc @cloud-fan

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 1]
```diff
@@ -52,9 +52,9 @@ grammar SqlBase;
   }
 
   /**
-   * When true, the behavior of keywords follows ANSI SQL standard.
+   * When true, ANSI SQL parsing mode is enabled.
    */
-  public boolean SQL_standard_keyword_behavior = false;
```
Comment (cloud-fan) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: I think `SQL_standard_keyword_behavior` is better. Can we keep it? ANSI is too vague here.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 2]
```diff
@@ -919,12 +919,6 @@ qualifiedNameList
     : qualifiedName (',' qualifiedName)*
     ;
 
-functionName
```
Comment (cloud-fan) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: this is for ANSI-compliant, not pgsql dialect, we should keep it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 3]
```diff
@@ -860,14 +853,12 @@ object TypeCoercion {
       case Add(l @ DateType(), r @ IntegerType()) => DateAdd(l, r)
       case Add(l @ IntegerType(), r @ DateType()) => DateAdd(r, l)
       case Subtract(l @ DateType(), r @ IntegerType()) => DateSub(l, r)
-      case Subtract(l @ DateType(), r @ DateType()) =>
-        if (SQLConf.get.usePostgreSQLDialect) DateDiff(l, r) else SubtractDates(l, r)
-      case Subtract(l @ TimestampType(), r @ TimestampType()) =>
-        SubtractTimestamps(l, r)
+      case Subtract(l @ DateType(), r @ DateType()) => DateDiff(l, r)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: This is not a simple revert. We want to create `SubtractDates` by default, and we need to create a legacy config since the pgsql dialect is gone.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 4]
```diff
@@ -860,14 +853,12 @@ object TypeCoercion {
...
+      case Subtract(l @ TimestampType(), r @ TimestampType()) => TimestampDiff(l, r)
       case Subtract(l @ TimestampType(), r @ DateType()) =>
-        SubtractTimestamps(l, Cast(r, TimestampType))
+        TimestampDiff(l, Cast(r, TimestampType))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: we should keep it. It's nothing to do with pgsql dialect.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 5]
```diff
@@ -2126,7 +2126,7 @@ case class DatePart(field: Expression, source: Expression, child: Expression)
  * is set to 0 and the `microseconds` field is initialized to the microsecond difference
  * between the given timestamps.
  */
-case class SubtractTimestamps(endTimestamp: Expression, startTimestamp: Expression)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: ditto, keep it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 6]
```diff
@@ -2143,25 +2143,3 @@ case class SubtractTimestamps(endTimestamp: Expression, startTimestamp: Expressi
       s"new org.apache.spark.unsafe.types.CalendarInterval(0, 0, $end - $start)")
   }
 }
-
-/**
- * Returns the interval from the `left` date (inclusive) to the `right` date (exclusive).
- */
-case class SubtractDates(left: Expression, right: Expression)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala [hunk 7]
```diff
@@ -65,15 +65,11 @@ object StringUtils extends Logging {
     "(?s)" + out.result() // (?s) enables dotall mode, causing "." to match new lines
   }
 
-  private[this] val trueStrings =
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala]: unnecessary change.

**Comment by cloud-fan** [comment]
Note that, we want to revert pgsql dialect, but not all pgsql compatibility features. A simple policy: if a behavior is not protected by `conf.usePostgreSQLDialect`, it should not be reverted.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 3 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Yes, keep the #26112

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 4 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Ditto

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 1 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Yes, keep the #26584

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 2 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala (hunk 5 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: Yes

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala (hunk 7 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala]: Thanks, done in 6d978c6.

**Comment by xuanyuanking** [comment]
@cloud-fan Thanks for the guidance, it takes me some time to reprocess all the related commits.
The last change keeps the date subtraction and ANSI-compliant support. I also changed the PR description correspondingly.

File: docs/sql-keywords.md [hunk 8]
```diff
@@ -19,7 +19,7 @@ license: |
   limitations under the License.
 ---
 
-When `spark.sql.dialect=PostgreSQL` or keep default `spark.sql.dialect=Spark` with setting `spark.sql.dialect.spark.ansi.enabled` to true, Spark SQL will use the ANSI mode parser.
+When `spark.sql.dialect.spark.ansi.enabled` is true, Spark SQL will use the ANSI mode parser.
```
Comment (cloud-fan) [inline docs/sql-keywords.md]: I don't think we can do a clean revert, we need to check the diff of this PR and see if they make sense.

Here we should use `spark.sql.ansi.enabled`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 9]
```diff
@@ -287,7 +286,7 @@ class Analyzer(
           case (_, CalendarIntervalType) => Cast(TimeSub(l, r), l.dataType)
           case (TimestampType, _) => SubtractTimestamps(l, r)
           case (_, TimestampType) => SubtractTimestamps(l, r)
-          case (_, DateType) => if (conf.usePostgreSQLDialect) {
+          case (_, DateType) => if (conf.ansiEnabled) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Actually `SubtractDates` is the ansi behavior.

Since `date - date` is newly added in 3.0, we can just change this to `case (_, DateType) => SubtractDates(l, r)`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala [hunk 10]
```diff
@@ -1131,7 +1131,8 @@ class DateExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     checkEvaluation(SubtractDates(Literal(end), epochDate),
       IntervalUtils.stringToInterval(UTF8String.fromString("interval 49 years 9 months 4 days")))
     checkEvaluation(SubtractDates(epochDate, Literal(end)),
-      IntervalUtils.stringToInterval(UTF8String.fromString("interval -49 years -9 months -4 days")))
+      IntervalUtils.stringToInterval(UTF8String.fromString(
+        "interval -49 years -9 months -4 days")))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala]: uncessary change.

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/boolean.sql [hunk 11]
```diff
@@ -22,6 +22,7 @@ SELECT false AS `false`;
 
 SELECT boolean('t') AS true;
 
+-- [SPARK-27931] Trim the string when cast string type to boolean type
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/postgreSQL/boolean.sql]: we do trim. Do we need to keep the JIRA ID?

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/boolean.sql [hunk 12]
```diff
@@ -98,7 +101,7 @@ SELECT boolean('f') <= boolean('t') AS true;
 
 -- explicit casts to/from text
 SELECT boolean(string('TrUe')) AS true, boolean(string('fAlse')) AS `false`;
-
+-- [SPARK-27931] Trim the string when cast to boolean type
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/postgreSQL/boolean.sql]: ditto

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/int2.sql [hunk 13]
```diff
@@ -92,9 +92,11 @@ WHERE f1 > -32767;
 
 SELECT '' AS five, i.f1, i.f1 - int('2') AS x FROM INT2_TBL i;
 
-SELECT '' AS five, i.f1, i.f1 / smallint('2') AS x FROM INT2_TBL i;
+-- PostgreSQL `/` is the same with Spark `div` since SPARK-2659.
+SELECT '' AS five, i.f1, i.f1 div smallint('2') AS x FROM INT2_TBL i;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/postgreSQL/int2.sql]: let's keep the tests the same as pgsql. we should still use `/`.

File: docs/sql-keywords.md (hunk 8 above)
Comment (xuanyuanking) [reply docs/sql-keywords.md]: Yes, my mistake for missing here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 9 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Copy that, then we need to change the postgreSQL/date.sql.out correspondingly. Done in 17a3c77.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala (hunk 10 above)
Comment (xuanyuanking) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala]: Thanks, done in 17a3c77.

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/boolean.sql (hunk 11 above)
Comment (xuanyuanking) [reply sql/core/src/test/resources/sql-tests/inputs/postgreSQL/boolean.sql]: This is introduced by revert, delete in next commit.

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/int2.sql (hunk 13 above)
Comment (xuanyuanking) [reply sql/core/src/test/resources/sql-tests/inputs/postgreSQL/int2.sql]: Copy that, also changed the test result in 17a3c77.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by xuanyuanking** [comment]
Thanks Wenchen!

**Comment by dongjoon-hyun** [comment]
Thank you, @xuanyuanking and @cloud-fan .

Also, cc to the original authors, too. Although we removed these, we are thankful for your original contributions. Sorry again.
- @wangyum 
- @younggyuchun 
- @gengliangwang 
- @MaxGekk 
- @Ngone51
